경북대 BE_신성민 2주차 과제 (2단계, 3단계)

anaconda77 opened the pull request:
안녕하세요 경북대 BE 신성민 학생입니다. 2주차 과제 step2/step3 제출드립니다.

## 과제 수행 중에 궁금했던 점
- 저번 step1 수행 시에 도메인 객체가 하는 일이 부실하다는 피드백을 해주셨습니다. 그래서 이번 step2를 수행할때 유저 로그인 시 이메일&패스워드가 일치하는지 여부를 검증하는 메소드를 유저 모델 안에 작성하였습니다.
` public boolean login(String email, String password) {
        return this.email.equals(email) && this.password.equals(password);
    }`


  컨트롤러에서 DAO를 호출하여 유저 정보를 가져온 후 Request에 포함된 아이디&패스워드와 서로 일치하는지를 비교하는 작업을 합니다.
  `Member loginedMember = memberDao.getMemberByEmail(loginRequest.email());
          if(!loginedMember.login(loginRequest.email(), loginRequest.password())) {
              throw new LoginErrorException();
          }`


  이렇게 되면 메소드를 호출할 때 파라미터로 유저의 아이디와 패스워드를 넘겨주게 되는데, 가장 바깥 영역인 컨트롤러단에서 유저 개인 정보를 파라미터로 넘겨주는게 좋은 방식은 아닌 것 같다는 생각이 듭니다.. 
아직은 Service 계층이 도입되지 않아서 비즈니스 로직을 호출하기 위해서는 어쩔 수 없이 이러한 과정을 수행해야되는 것 같은데, 좀 더 나은 방안이 있을까요?

- 회원가입/로그인 시 jwt 토큰을 생성해서 응답 헤더에 넣어주는 작업을 현재 컨트롤러에서 하고 있는데, 해당 작업이 컨트롤러의 역할과 책임에 맞지 않다는 생각이 듭니다.
 따라서 해당 작업을 수행해줄 별도의 객체가 있어야된다는 생각이 드는데, 찾아보니 Spring Security에서는 SuccessHandler라는 것을 통해 로그인 관련 로직을 처리하는 것 같습니다. 
그러나 이번 과제에서는 Spring Security 없이 jwt 인증 시스템 구현을 요구하셨기 때문에, 해당 핸들러를 사용하지는 못합니다. 
현재 시점에선 컨트롤러에서 직접 토큰 생성 작업을 수행하게 하는 것이 최선일까요? 아님 다른 방법이 있는지 궁금합니다.

피드백 해주셔서 감사합니다:)

anaconda77 committed:
refactor: DTO 이름 변경

anaconda77 committed:
refactor: DTO record 클래스로 변경

anaconda77 committed:
refactor: print문 삭제

anaconda77 committed:
refactor: 상품 도메인 내 수정 메소드 추가

anaconda77 committed:
test: valid/invalid 테스팅 메서드 분리

anaconda77 committed:
docs: README.md step3 구현할기능목록 작성

anaconda77 committed:
feat: Member 도메인 객체 생성

anaconda77 committed:
feat: Member 테이블 스키마/데모 데이터  작성

anaconda77 committed:
feat: MemberDao 생성

anaconda77 committed:
feat: 로그인 폼 객체 생성

anaconda77 committed:
feat: 이메일(아이디) 중복 불가 스키마 수정

anaconda77 committed:
"feat: MemberDao 수정"

유저 변경/삭제 기능은 필요없으므로 삭제
이메일 중복/조회 데이터 없음 예외 명시

anaconda77 committed:
feat: 유저 가입 request 객체 추가

anaconda77 committed:
refactor: ProductExceptionHandler 명칭 변경

MemberRequest에도 활용가능하므로 InputExceptionHandler로 명칭 변경

anaconda77 committed:
"refactor: ProductException 변경"

ProductException과 MemberException이 수행하는 역할이 동일하므로 MemberException을 별도로 만들지 않고, 둘다 사용할 수 있도록 InputException으로 네이밍 변경

anaconda77 committed:
feat: 회원 가입 요청/응답 dto 객체 생성

anaconda77 committed:
feat: 회원 가입 예외 핸들러 구현

anaconda77 committed:
feat: 회원 가입 기능 구현

anaconda77 committed:
README.md 업데이트

anaconda77 committed:
refactor: 로그인 커스텀 예외 추가

anaconda77 committed:
feat: 회원 로그인 기능 구현

anaconda77 committed:
docs: README.md 업데이트

anaconda77 committed:
refactor: 로그인 관련 응답 상태 코드 변경

400->403으로 변경

anaconda77 committed:
feat: WebConfig 작성

anaconda77 committed:
feat: AuthInterceptor 작성

anaconda77 committed:
feat: 유저 회원가입/로그인 시 session 전달 기능 구현

anaconda77 committed:
fix: webconfig 애노테이션 명시

anaconda77 committed:
feat: 인증 실패에 대한 응답코드 반환

anaconda77 committed:
docs: README.md 업데이트

anaconda77 committed:
chore: 컨트롤러 위치 변경

anaconda77 committed:
refactor: MemberDao 예외 throws 명시 제거

anaconda77 committed:
refactor: 로그인 실패 응답 상태코드 변경

anaconda77 committed:
refactor: 예외 핸들링에 불필요한 ResponseEntity<> 제거

anaconda77 committed:
docs: REAMD.md 요구사항 변경

과제 요구사항에 맞춰 인증 방식을 세션에서 JWT로 변경함

anaconda77 committed:
feat: jwt 초기 설정

application.properties 파일에 시크릿 키 등 민감한 정보가 포함되어 있으므로, gitignore에 추가함

anaconda77 committed:
feat: 토큰 생성기 구현

anaconda77 committed:
feat: 토큰 검증 인터셉터 구현

anaconda77 committed:
feat: 회원가입/로그인 시 세션 대신 토큰 반환하도록 메소드 수정

anaconda77 committed:
fix: 인터셉터 예외 처리 에러 해결

anaconda77 committed:
feat: 토큰에 유저 정보 모두 포함

anaconda77 committed:
feat: LoginMemberArgumentResolver 구현

anaconda77 committed:
feat: MemberDao insert() 리턴값으로 Member 반환

anaconda77 committed:
docs: README.md step3 작성

anaconda77 committed:
feat: WishProducts 테이블/데모 데이터 작성

anaconda77 committed:
feat: WishProduct 도메인 객체 생성

anaconda77 committed:
feat: 위시 리스트 조회 기능 구현

anaconda77 committed:
feat: 위시 리스트 추가 기능 구현

anaconda77 committed:
feat: 위시 리스트 추가 dto 생성

anaconda77 committed:
docs: README.md 업데이트

anaconda77 committed:
refactor: wishproducts 스키마 수정

anaconda77 committed:
refactor: 모델/dao 변경

anaconda77 committed:
feat: WishProducts pk 설정

anaconda77 committed:
refactor: 유저 관련 예외 분리

anaconda77 committed:
feat: 위시 리스트 추가 기능 구현

anaconda77 committed:
feat: 위시 리스트 삭제 기능 구현

anaconda77 committed:
fix: 위시 리스트 삭제 예외 처리

anaconda77 committed:
refactor: 예외 메시지를  DAO에서 생성

anaconda77 committed:
refactor: 패키지 구조 리팩터링

lsm7179 commented on .gitignore:
```diff
@@ -25,6 +25,7 @@ bin/
 out/
 !**/src/main/**/out/
 !**/src/test/**/out/
+application.properties
```
프로퍼티를 숨기고 싶다면 
`git Submodules` 도 있습니다~
https://git-scm.com/book/en/v2/Git-Tools-Submodules
https://devocean.sk.com/blog/techBoardDetail.do?ID=165172&boardType=techBlog

anaconda77 replied:
아하 이런것도 있군요..프로젝트에서 보안상 민감한 파일들은 프라이빗 리포지토리에서 따로 관리하면 될 것 같네요!

lsm7179 commented on src/main/java/gift/auth/AuthInterceptor.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response,
+        Object handler) throws Exception {
+
+        String token = tokenProvider.extractJwtTokenFromHeader(request);
+
+        //토큰이 존재하지 않음
+        if(token == null) {
+            response.setStatus(HttpStatus.UNAUTHORIZED.value());
+            return false;
+        }
```
👍

lsm7179 commented on src/main/java/gift/auth/JwtTokenProvider.java:
```diff
@@ -0,0 +1,69 @@
+            .claim("member_id", member.getId())
+            .setIssuedAt(new Date(System.currentTimeMillis()))
+            .setExpiration(new Date(System.currentTimeMillis() + expirationInSeconds))
+            .signWith(HS512, secretKey)
```
'signWith(io.jsonwebtoken.SignatureAlgorithm, java.lang.String)' is deprecated 

deprecated되었네요.
수정이 필요해보여요~

lsm7179 commented on README.md:
```diff
@@ -32,4 +32,63 @@
+  - th:errors/errorclass를 활용하여 에러 출력
+
+    
+## 2단계
+### 유저 인증을 통한 기능 사용 권한 검증
```
요구사항과 구현방법을 상세히 설명해주셨네요~👍💯

lsm7179 commented on src/main/java/gift/auth/JwtTokenProvider.java:
```diff
@@ -0,0 +1,69 @@
+        try {
+            return Jwts
+                .parserBuilder()
+                .setSigningKey(secretKey)
```
.parser().setSigningKey도
deprecated되었네요.
수정이 필요해보여요~

lsm7179 commented on src/main/java/gift/config/WebConfig.java:
```diff
@@ -0,0 +1,35 @@
+import org.springframework.web.servlet.config.annotation.WebMvcConfigurer;
+
+@Configuration
+public class WebConfig implements WebMvcConfigurer {
```
잘 설정해주었지만 webconfig가 많아질수록 해당 클래스명으로 어떤 작업을 하는지 추측하기 어려울거 같아요.
loginWebConfig 처럼 로그인 관련된 작업을 하는 클래스다 라고 명명하면 어떨까요?

lsm7179 commented on src/main/java/gift/controller/LoginController.java:
```diff
@@ -0,0 +1,62 @@
+        Member member = new Member(joinRequest.email(), joinRequest.password());
+        Member joinedMember = memberDao.insertMember(member);
+
+        response.setHeader("Authorization",jwtTokenProvider.generateToken(joinedMember));
+
```
회원가입하면 자동으로 로그인까지 구현해주셨군요.

lsm7179 commented on src/main/java/gift/auth/JwtTokenProvider.java:
```diff
@@ -0,0 +1,69 @@
+        return null;
+    }
+
+    public boolean isExpired(String token) {
```
사용하지 않는 코드는 지워주세요~

lsm7179 commented on src/main/java/gift/controller/LoginController.java:
```diff
@@ -0,0 +1,62 @@
+            throw new InputException(bindingResult.getAllErrors());
+        }
+
+        Member loginedMember = memberDao.getMemberByEmail(loginRequest.email());
+        if(!loginedMember.login(loginRequest.email(), loginRequest.password())) {
+            throw new LoginErrorException();
+        }
```
service계층이 있었으면 좋겠습니다.
컨트롤러가 이 모든 행위를 담당해야될까요?
서비스에서 로그인 시도해봐 라고 호출해보세요~

lsm7179 commented on src/main/java/gift/controller/LoginController.java:
```diff
@@ -0,0 +1,62 @@
+            throw new InputException(bindingResult.getAllErrors());
+        }
+
+        Member member = new Member(joinRequest.email(), joinRequest.password());
+        Member joinedMember = memberDao.insertMember(member);
```
service계층이 있었으면 좋겠습니다.
컨트롤러가 이 모든 행위를 담당해야될까요?
서비스에서 회원가입 해봐 라고 호출해보세요~

lsm7179 commented on src/main/java/gift/response/ProductResponse.java:
```diff
@@ -1,10 +1,10 @@
 import gift.model.Product;
 
-public record ProductResponseDto(Long id, String name, int price, String imageUrl) {
+public record ProductResponse(Long id, String name, int price, String imageUrl) {
```
👍

lsm7179 commented on src/main/java/gift/controller/ProductViewController.java:
```diff
@@ -60,14 +60,11 @@ public String editForm(@RequestParam Long id, Model model) {
-        updatedProduct.setName(form.getName());
-        updatedProduct.setPrice(form.getPrice());
-        updatedProduct.setImageUrl(form.getImageUrl());
+        updatedProduct.update(form.getName(), form.getPrice(), form.getImageUrl());
```
도메인객체로 넘겨서 잘해주셨네요~ loginController에서 말했던 거처럼 Service계층에서 책임을 가질수있게 해보세요~👍

lsm7179 commented on src/main/java/gift/controller/WishListApiController.java:
```diff
@@ -0,0 +1,73 @@
+    public ResponseEntity<List<ProductResponse>> getWishList(@LoginMember LoginMemberDto memberDto) {
+
+        List<Product> wishlist = wishProductDao.findAll(memberDto.id());
+        if (wishlist.isEmpty()) {
+            return new ResponseEntity<>(HttpStatus.OK);
+        }
+
+        List<ProductResponse> dtoList = wishlist.stream()
+            .map(ProductResponse::new)
+            .toList();
```
클라이언트에서 ProductResponse를 받을때 빈값이라도 내려주는것이 좋습니다.
왜 그런지 생각해보세요~

anaconda77 replied:
프론트에서 조회 결과를 리스트로 받아오게 설계하기 때문에, 결과가 없더라도 빈 리스트 구조를 유지하는게 통일성이 맞을 것 같습니다..!

lsm7179 commented on src/main/java/gift/exception/MemberExceptionHandler.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+@RestControllerAdvice
+public class MemberExceptionHandler {
+
+    @ResponseStatus(HttpStatus.FORBIDDEN)
+    @ExceptionHandler(MemberException.class)
+    public ErrorResult joinExHandle(MemberException e) {
+        return new ErrorResult("회원가입 에러", e.getMessage());
+    }
+
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    @ExceptionHandler({LoginErrorException.class})
+    public ErrorResult notFoundUserExHandle(LoginErrorException e) {
+        return new ErrorResult("로그인 에러", e.getMessage());
+    }
```
👍

lsm7179 commented on src/main/java/gift/model/Product.java:
```diff
@@ -25,6 +25,12 @@ public Product() {
 
     }
 
+    public void update(String name, Integer price, String imageUrl) {
+        this.name = name;
+        this.price = price;
+        this.imageUrl = imageUrl;
+    }
+
```
👍

lsm7179 commented:
> 안녕하세요 경북대 BE 신성민 학생입니다. 2주차 과제 step2/step3 제출드립니다.
> 
> ## 과제 수행 중에 궁금했던 점
>   컨트롤러에서 DAO를 호출하여 유저 정보를 가져온 후 Request에 포함된 아이디&패스워드와 서로 일치하는지를 비교하는 작업을 합니다.
>   `Member loginedMember = memberDao.getMemberByEmail(loginRequest.email()); if(!loginedMember.login(loginRequest.email(), loginRequest.password())) { throw new LoginErrorException(); }`
>   이렇게 되면 메소드를 호출할 때 파라미터로 유저의 아이디와 패스워드를 넘겨주게 되는데, 가장 바깥 영역인 컨트롤러단에서 유저 개인 정보를 파라미터로 넘겨주는게 좋은 방식은 아닌 것 같다는 생각이 듭니다..

좋은 고민입니다.
Service 계층을 이용해서 제어해주면 어떨까요?

>   아직은 Service 계층이 도입되지 않아서 비즈니스 로직을 호출하기 위해서는 어쩔 수 없이 이러한 과정을 수행해야되는 것 같은데, 좀 더 나은 방안이 있을까요?

Service계층을 도입하면 좋을거 같습니다~!

> * 회원가입/로그인 시 jwt 토큰을 생성해서 응답 헤더에 넣어주는 작업을 현재 컨트롤러에서 하고 있는데, 해당 작업이 컨트롤러의 역할과 책임에 맞지 않다는 생각이 듭니다.
>   따라서 해당 작업을 수행해줄 별도의 객체가 있어야된다는 생각이 드는데, 찾아보니 Spring Security에서는 SuccessHandler라는 것을 통해 로그인 관련 로직을 처리하는 것 같습니다.
>   그러나 이번 과제에서는 Spring Security 없이 jwt 인증 시스템 구현을 요구하셨기 때문에, 해당 핸들러를 사용하지는 못합니다.
>   현재 시점에선 컨트롤러에서 직접 토큰 생성 작업을 수행하게 하는 것이 최선일까요? 아님 다른 방법이 있는지 궁금합니다.

Token생성 서비스 계층으로 생성로직을 넘겨주면 좋을거 같아요. ControllerAdvice, Filter, interceptor 등을 활용하면 더 좋을거 같습니다~

lsm7179 reviewed (changes requested):
성민님 2단계3단계 잘 해주셨어요.
요구사항은 잘 충족했으나 몇가지만 고치면 좋을거 같아서 반영해보시고 다시 리뷰요청해주세요~

anaconda77 committed:
chore: git submodule 설정

application.properties에 있는 민감한 보안 사항을 숨기기 위해 private 레포지토리에 별도로 보관하고, submodule 설정으로 해당 파일을 프로젝트에 불러올 수 있게 함

anaconda77 committed:
refactor: jjwt 최신버전 반영

anaconda77 committed:
refactor: config 네임 변경

anaconda77 committed:
refactor: MemberService 도입

anaconda77 committed:
refactor: ProductService 도입

anaconda77 committed:
refactor: 상품/위시리스트 전체조회 응답값 리스트로 통일

anaconda77 committed:
refactor: JwtService 생성

lsm7179 commented on src/main/java/gift/auth/JwtTokenProvider.java:
```diff
@@ -0,0 +1,64 @@
+    public String extractJwtTokenFromHeader(HttpServletRequest request) {
+        String bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
+
+        if (bearerToken != null && bearerToken.startsWith("Bearer ")) {
+            return bearerToken.substring(7);
+        }
```
문자와 숫자가 어떤 뜻인지 알기 어려울수있어요.
상수로 만들어서 해보세요~

lsm7179 commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class MemberService {
```
👍 서비스 계층을 만드셨네요.

lsm7179 commented on src/main/java/gift/controller/WishListApiController.java:
```diff
@@ -0,0 +1,74 @@
+    @GetMapping("/api/wishlist")
+    public ResponseEntity<List<ProductResponse>> getWishList(@LoginMember LoginMemberDto memberDto) {
+
+        List<Product> wishlist = wishProductDao.findAll(memberDto.id());
+        if (wishlist.isEmpty()) {
+            return new ResponseEntity<>( new ArrayList<>(), HttpStatus.OK);
+        }
```
wishlist가 이미 빈값인데 새로 빈 리스트를 생성해서 내려주는 이유가 있나요?

anaconda77 replied:
앗 기본적인 실수를 했네요ㅜ피드백 주신 부분 다음주차에 수정해서 올려놓도록 하겠습니다!

lsm7179 reviewed (approved):
성민님
수고하셨습니다.
요구사항 잘 구현해주셨어요.👍
머지하겠습니다~

궁금하거나 어려운거 있으시면 편하게 DM 주세요.
